File: leetcode/daily-code/10-22.js
var partitionLabels = function(S) {
  var pos = new Array(26);
  const cpa = 'a'.charCodeAt(0);
  for (var i = 0; i < S.length; ++i) {
    pos[S.charCodeAt(i) - cpa] = i;
  }

  var ret = new Array();
  var start = 0;
  var end = 0;
  for (var i = 0; i < S.length; ++i) {
    end = Math.max(end, pos[S.charCodeAt(i) - cpa]);
    if (end == i) {
      ret.push(end - start + 1);
      start = end + 1;
    }
  }
  return ret;
};

File: leetcode/daily-code/10-22.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        vector<int> ret;
        int pos[26];
        for(auto i = 0; i< S.size(); ++i){
            pos[S[i]-'a'] = i;
        }
        int start = 0;
        int end = 0;
        for(auto i= 0; i< S.size(); ++i){
            end = std::max(pos[S[i]-'a'],end);
            if(end == i){
                ret.push_back(end-start+1);
                start = end+1;
                i = end;
            }

        }
        return ret;
    }
};